skrifa/src/meta/strings.rs

```diff
@@ -0,0 +1,683 @@
+//! Strings describing font names and other metadata in multiple languages.
+//!
+//! This provides higher level interfaces for accessing the data in the
+//! OpenType [name](https://learn.microsoft.com/en-us/typography/opentype/spec/name)
+//! table.
+//!
+//! # Example
+//! The following function will print all localized strings in a font:
+//! ```
+//! use skrifa::MetadataProvider;
+//!
+//! fn print_all_strings<'a>(font: &impl MetadataProvider<'a>) {
+//!     for id in font.string_ids() {
+//!         println!("[{:?}]", id);
+//!         for string in font.localized_strings(id) {
+//!             println!("{:?} {}", string.language(), string.to_string());
+//!         }
+//!     }
+//! }
+//! ```
+
+use read_fonts::{
+    tables::name::{CharIter, Name, NameRecord, NameString},
+    TableProvider,
+};
+
+use core::fmt;
+
+#[doc(inline)]
+pub use read_fonts::types::NameId as StringId;
+
+#[derive(Clone)]
+/// Iterator over the characters of a string.
+pub struct Chars<'a> {
+    inner: Option<CharIter<'a>>,
+}
+
+impl<'a> Iterator for Chars<'a> {
+    type Item = char;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner.as_mut()?.next()
+    }
+}
+
+/// Iterator over the collection of available string identifiers in a font.
+pub struct StringIds<'a> {
+    records: core::slice::Iter<'a, NameRecord>,
+    last_id: Option<StringId>,
+}
+
+impl<'a> StringIds<'a> {
+    /// Creates a new string identifier iterator from the given font.
+    pub fn new(font: &impl TableProvider<'a>) -> Self {
+        let name = font.name().ok();
+        let records = name
+            .as_ref()
+            .map(|name| name.name_record().iter())
+            .unwrap_or([].iter());
+        Self {
+            records,
+            last_id: None,
+        }
+    }
```

Comment:
This feels like a minor mixing of concerns to me, although I'm not sure it's that important?

Basically: if we don't have a name table, it would make sense to me to handle that by returning an `Option<StringIds<'a>>` from the method on metadata provider, and I would maybe make this constructor take a name table directly? To put it another way, this method is silently handling an error condition in a way that makes it difficult to identify. There might be a good reason for this choice, but it does jump out at me.

Reply:
This is mostly about ergonomics. The intent, specifically with metadata, was to not surface these types of things. My expectation is that for anyone consuming this library, the logic for dealing with `Option<StringIds<'a>>` and an empty iterator would be exactly the same so I'd like to not impose useless boilerplate.

Generally, I'd expect implementations that require a minimum set of (valid?) tables to perform that validation at font loading time using the more fine-grained error reporting provided by `read-fonts`.

Reply:
I think what I'd like to do here is document this behavior and add a method that reports the set of supported metadata that is available in the font based on the presence of required tables. This lets client code determine if a font is considered usable for its purposes at load time rather than having to deal with errors at lower levels in the stack. This feels better to me for integration in full scale text rendering systems but I'd be happy to discuss further in the forthcoming PR.